discord/app/cache.py

```diff
@@ -0,0 +1,306 @@
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from collections import OrderedDict, deque
+from typing import Deque, Protocol
+
+from discord.app.state import ConnectionState
+from discord.message import Message
+
+from ..abc import PrivateChannel
+from ..channel import DMChannel
+from ..emoji import AppEmoji, GuildEmoji
+from ..guild import Guild
+from ..poll import Poll
+from ..sticker import GuildSticker, Sticker
+from ..ui.modal import Modal, ModalStore
+from ..ui.view import View, ViewStore
+from ..user import User
+from ..types.user import User as UserPayload
+from ..types.emoji import Emoji as EmojiPayload
+from ..types.sticker import GuildSticker as GuildStickerPayload
+from ..types.channel import DMChannel as DMChannelPayload
+
+class Cache(Protocol):
+    # users
+    async def get_all_users(self) -> list[User]:
+        ...
+
+    async def store_user(self, payload: UserPayload) -> User:
+        ...
+
+    async def delete_user(self, user_id: int) -> None:
+        ...
+
+    async def get_user(self, user_id: int) -> User | None:
+        ...
+
+    # stickers
+
+    async def get_all_stickers(self) -> list[GuildSticker]:
+        ...
+
+    async def get_sticker(self, sticker_id: int) -> GuildSticker:
+        ...
+
+    async def store_sticker(self, guild: Guild, data: GuildStickerPayload) -> GuildSticker:
+        ...
+
+    # interactions
+
+    async def store_view(self, view: View, message_id: int | None) -> None:
+        ...
+
+    async def delete_view_on(self, message_id: int) -> None:
+        ...
+
+    async def get_all_views(self) -> list[View]:
+        ...
+
+    async def store_modal(self, modal: Modal) -> None:
+        ...
+
+    async def get_all_modals(self) -> list[Modal]:
+        ...
+
+    # guilds
+
+    async def get_all_guilds(self) -> list[Guild]:
+        ...
+
+    async def get_guild(self, id: int) -> Guild:
+        ...
+
+    async def add_guild(self, guild: Guild) -> None:
+        ...
+
+    async def delete_guild(self, guild: Guild) -> None:
+        ...
+
+    # emojis
+
+    async def store_guild_emoji(self, guild: Guild, data: EmojiPayload) -> GuildEmoji:
+        ...
+
+    async def store_app_emoji(
+        self, application_id: int, data: EmojiPayload
+    ) -> AppEmoji:
+        ...
+
+    async def get_all_emojis(self) -> list[GuildEmoji | AppEmoji]:
+        ...
+
+    async def get_emoji(self, emoji_id: int | None) -> GuildEmoji | AppEmoji | None:
+        ...
+
+    async def delete_emoji(self, emoji: GuildEmoji | AppEmoji) -> None:
+        ...
+
+    # polls
+
+    async def get_all_polls(self) -> list[Poll]:
```

Comment:
The fact that every cache would have to implement all of these might be not that practical - we can keep this and have an even lower level thing that could also be overriden that would just store key value pairs and use that in the default cache as a dict

Reply:
That's what `MemoryCache` is for as well. You subclass that class if you have some data you don't wanna store outside of memory but still wanna store others in memory

Reply:
Fair makes sense. I'm still not sure of some things but I see your point